File: ref/code/L15/U2021decs2.java
package com.deltacoding.usaco.s2021_2022.dec2021.silver;

import java.io.*;
import java.util.*;

public class U2021decs2 {
	static int[] f;
	
	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		int T = Integer.parseInt(in.readLine());
		
		for(int i = 0; i < T; i++) {
			singledata(in);
		}

	}
	
	static void singledata(BufferedReader in) throws IOException {
        StringTokenizer st = new StringTokenizer(in.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
		f = new int[n + 1];
		
		ArrayList<Integer>[] lists = new ArrayList[n + 1];
		for(int i = 1; i <= n; i++) {
			lists[i] = new ArrayList<>();
			f[i] = i;
		}
		for(int i = 0; i < m; i++) {
			st = new StringTokenizer(in.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			merge(a, b);
		}

		int f1 = find(1);
		int fn = find(n);
		if(f1 == fn) {
			System.out.println(0);
			return;
		}

		for(int i = 1; i<=n; i++) {
			lists[find(i)].add(i);
		}
	
		int[] arrf1 = new int[lists[f1].size()];
		for(int i=0; i<arrf1.length; i++) {
			arrf1[i] = lists[f1].get(i);
		}
		int[] arrfn = new int[lists[fn].size()];
		for(int i=0; i<arrfn.length; i++) {
			arrfn[i] = lists[fn].get(i);
		}
		
		long ans = Long.MAX_VALUE;
		for(int i = 2; i <= n-1; i++) {
			if(find(i) != i) {
				continue;
			}
			long min1 = Integer.MAX_VALUE, min2 = Integer.MAX_VALUE; // 确保min1*min1在long范围内
			for(int node : lists[i]) { // 枚举该连通块中每一个点
				int pos = binarySearch(arrf1, node); // 找大于等于且最接近node的数的位置
				if(pos < arrf1.length) {
					min1 = Math.min(min1, arrf1[pos] - node);
				}
				if(pos > 0) {
					min1 = Math.min(min1, node - arrf1[pos - 1]);
				}

				pos = binarySearch(arrfn, node); // 找大于等于且最接近node的数的位置
				if(pos < arrfn.length) {
					min2 = Math.min(min2, arrfn[pos] - node);
				}
				if(pos > 0) {
					min2 = Math.min(min2, node - arrfn[pos - 1]);
				}
			}
			
			ans = Math.min(ans, min1 * min1 + min2 * min2);
		}
		System.out.println(ans);
	}
	
	static int binarySearch(int[] arr, int key) {
		int l = 0, r = arr.length - 1;
		int ans = arr.length;
		while(l <= r) {
			int mid = (l + r) / 2;
			if(arr[mid] >= key) {
				ans = mid;
				r = mid - 1;
			} else {
				l = mid + 1;
			}
		}
		return ans;
	}

	static void merge(int x, int y) {
		int fx = find(x);
		int fy = find(y);
		if(fx != fy) {
			f[fy] = fx;
		}
	}
	
	static int find(int u) {
		if (f[u] == u) {
			return u;
		}

		f[u] = find(f[u]); // 路径压缩

		return f[u];
	}
}

File: ref/code/L15/U2021decs2.cpp
#include <bits/stdc++.h>
using namespace std;
int n, m, t;
int f[100005];
vector<int> v[100005];

int find(int x) {
    if (x == f[x]) {
        return x;
    }
    f[x] = find(f[x]);
    return f[x];
}

void merge(int u, int v) {
    int fu = find(u);
    int fv = find(v);
    if (fu != fv) {
        f[fu] = fv;
    }
}

/*
先使用并查集维护好点与点之间的连接情况
把每个点存到它对应根节点的vector或arraylist中
再把1和n所在连通块的节点分别存好
枚举用哪一个联通块把1和n所在的联通块连接起来（可以利用二分查找）
*/
int main() {
    cin >> t;
    for (int i = 1; i <= t; i++) {
        cin >> n >> m;
        for (int j = 1; j <= n; j++) {
            f[j] = j;
        }
        for (int j = 1; j <= m; j++) {
            int a, b;
            cin >> a >> b;
            merge(a, b);
        }
        int f1 = find(1), fn = find(n);
        if (f1 == fn) {
            cout << 0 << endl;
            continue;
        }

        for (int j = 1; j <= n; j++) v[j].clear();
        for (int j = 1; j <= n; j++) v[find(j)].push_back(j);

        long long ans = LONG_MAX;
        for (int j = 2; j < n; j++) {
            if (find(j) != j) {
                continue;
            }
            int min1 = n, min2 = n;
            for (int k : v[j]) {  // 枚举j所在连通块里的所有点
                int pos = lower_bound(v[f1].begin(), v[f1].end(), k) - v[f1].begin();
                if (pos < v[f1].size()) {
                    // 把k号点和1号点所在连通块中大于等于k大且最接近它的点连接
                    min1 = min(min1, v[f1][pos] - k);
                }
                if (pos > 0) {
                    // 把k号点和1号点所在连通块中比k小且最接近它的点连接
                    min1 = min(min1, k - v[f1][pos - 1]);
                }

                pos = lower_bound(v[fn].begin(), v[fn].end(), k) - v[fn].begin();
                if (pos < v[fn].size())
                    // 把k号点和n号点所在连通块中大于等于k大且最接近它的点连接
                    min2 = min(min2, v[fn][pos] - k);
                if (pos > 0) {
                    // 把k号点和n号点所在连通块中比k小且最接近它的点连接
                    min2 = min(min2, k - v[fn][pos - 1]);
                }
            }
            ans = min(ans, 1LL * min1 * min1 + 1LL * min2 * min2);
        }
        cout << ans << endl;
    }
    return 0;
}

/*
2
5 2
1 2
4 5
5 3
1 2
2 3
4 5
*/
